android-o/AutofillFramework/AutofillFramework/app/CustomVirtualView.cs

```diff
@@ -0,0 +1,189 @@
+		public static int sNextId;
+
+		List<Line> VirtualViewGroups = new List<Line>();
+		SparseArray<Item> VirtualViews = new SparseArray<Item>();
+		AutofillManager AutofillManager;
+
+    	Line FocusedLine;
+		Paint TextPaint;
+
+		Line UsernameLine;
+		Line PasswordLine;
+
+		public CustomVirtualView(Context context, IAttributeSet attrs) : base(context, attrs)
+		{
+			AutofillManager = (AutofillManager)context.GetSystemService(Class.FromType(typeof(AutofillManager)));
+			TextPaint = new Paint();
+			TextPaint.SetStyle(Style.Fill);
+			TextPaint.TextSize = TEXT_HEIGHT;
+			UsernameLine = AddLine("usernameField", context.GetString(Resource.String.username_label),
+						new string[] { AutofillHintUsername }, "         ", true);
+			PasswordLine = AddLine("passwordField", context.GetString(Resource.String.password_label),
+						new string[] { AutofillHintPassword }, "         ", false);
+		}
+
+		public override void Autofill(SparseArray values)
+		{
+			// User has just selected a Dataset from the list of autofill suggestions.
+			// The Dataset is comprised of a list of AutofillValues, with each AutofillValue meant
+			// to fill a specific autofillable view. Now we have to update the UI based on the
+			// AutofillValues in the list.
+			Log.Debug(LogTag, "autoFill(): " + values);
+			for (int i = 0; i < values.Size(); i++)
+			{
+				var id = values.KeyAt(i);
+				var value = (AutofillValue)values.ValueAt(i);
+				Item item = VirtualViews.Get(id);
+				if (item != null && item.Editable)
+				{
+					// Set the item's text to the text wrapped in the AutofillValue.
+					item.Text = value.TextValue;
+				}
+				else if (item == null)
+				{
+					Log.Warn(LogTag, "No item for id " + id);
+				}
+				else
+				{
+					Log.Warn(LogTag, "Item for id " + id + " is not editable: " + item);
+				}
+			}
+			PostInvalidate();
+		}
+
+		public override void OnProvideAutofillVirtualStructure(ViewStructure structure, AutofillFlags flags)
+		{
+			// Build a ViewStructure that will get passed to the AutofillService by the framework
+			// when it is time to find autofill suggestions.
+			structure.SetClassName(Class.Name);
+			var childrenSize = VirtualViews.Size();
+			Log.Debug(LogTag, "onProvideAutofillVirtualStructure(): flags = " + flags + ", items = "
+			          + childrenSize + ", extras: " + CommonUtil.BundleToString(structure.Extras));
+			var index = structure.AddChildCount(childrenSize);
+			// Traverse through the view hierarchy, including virtual child views. For each view, we
+			// need to set the relevant autofill metadata and add it to the ViewStructure.
+			for (int i = 0; i < childrenSize; i++)
+			{
+				Item item = VirtualViews.ValueAt(i);
+				Log.Debug(LogTag, "Adding new child at index " + index + ": " + item);
+				var child = structure.NewChild(index);
+				child.SetAutofillId(structure.AutofillId, item.Id);
+				child.SetAutofillHints(item.Hints);
+				child.SetAutofillType((int) item.Type); // Fix enumification
```

Comment:
This has been fixed in d15-4 so we can remove the cast and comment here now.

Reply:
Done